src/raster/array.jl

```diff
@@ -1,89 +1,106 @@
+import DiskArrays: eachchunk, GridChunks
+import DiskArrays
 const AllowedXY = Union{Integer,Colon,AbstractRange}
 const AllowedBand = Union{Integer,Colon,AbstractArray}
 
 
+# Define a RasterDataset type
+struct RasterDataset{T,DS} <: AbstractDiskArray{T,3}
+    ds::DS
+    size::Tuple{Int,Int,Int}
+end
+#Forward a few functions
+for f in (:getgeotransform, :nraster, :getband, :getproj,
+  :width, :height, :destroy)
+
+  eval(:($(f)(x::RasterDataset) = $(f)(x.ds)))
+end
+function RasterDataset(ds::AbstractDataset)
+    iszero(nraster(ds)) && throw(ArgumentError("The Dataset does not contain any raster bands"))
+    s = _common_size(ds)
+    RasterDataset{_dataset_type(ds), typeof(ds)}(ds,s)
+end
+function _dataset_type(ds::AbstractDataset)
+  alldatatypes = map(1:nraster(ds)) do i
+    b = getband(ds,i)
+    pixeltype(b)
+  end
+  reduce(promote_type,alldatatypes)
+end
+function _common_size(ds::AbstractDataset)
+  nr = nraster(ds)
+  allsizes = map(1:nr) do i
+    b = getband(ds,i)
+    size(b)
+  end
+  s = unique(allsizes)
+  length(s) == 1 || throw(DimensionMismatch("Can not coerce bands to single dataset, different sizes found"))
+  Int64.((s[1]...,nr))
+end
+getband(ds::RasterDataset,i) = getband(ds.ds,i)
+unsafe_readraster(args...;kwargs...)  = RasterDataset(unsafe_read(args...;kwargs...))
+readraster(s::String;kwargs...) = RasterDataset(read(s;kwargs...))
+function eachchunk(ds::RasterDataset)
+  subchunks = eachchunk(getband(ds,1))
+  GridChunks(ds,(subchunks.chunksize...,1))
+end
+DiskArrays.haschunks(::RasterDataset) = DiskArrays.Chunked()
+DiskArrays.haschunks(::AbstractRasterBand) = DiskArrays.Chunked()
 # AbstractRasterBand indexing
 
 Base.size(band::AbstractRasterBand) = width(band), height(band)
-Base.firstindex(band::AbstractRasterBand, d) = 1
-Base.lastindex(band::AbstractRasterBand, d) = size(band)[d]
+# Don't know if we need these, commenting out for the moment
```

Comment:
@rafaqz @visr are `Base.firstindex(...)` and `Base.lastindex(...)` still required? If they are, can we try to understand when they might be used?

Reply:
AFAIK they are used when users run A[begin:end,:] where `begin` gets replaced by `firstindex(A,1)` and `end` by `lastindex(A,1)`. However, since this is an `AbstractArray` now there are fallback methods implemented https://github.com/JuliaLang/julia/blob/master/base/abstractarray.jl#L303

Reply:
Yeah those  aren't needed now this is wrapped with `AbstractArray`.

Reply:
Ok if we don't need them, it's probably best to remove them altogether.